File: leetcode/46.go
package main

import "fmt"

func main() {
    nums := []int{1,2,3}
    ans := permute(nums)
    fmt.Println(ans)
}


func permute(nums []int) [][]int {
    // var res = [][]int{}
    res := [][]int{}
    record := []int{}
    ans := backtrack(nums, record, res)
    return ans
}

func backtrack(nums, track []int, ans [][]int) [][]int {
    size := len(nums)
    if size == len(track) {
        temp := make([]int, size)
        copy(temp, track)
        ans = append(ans, temp)
        return ans
    }
    for _, num := range nums {
        flag := 0
        for _, temp := range track {
            if temp == num {
                flag = 1
                break
            }
        }
        if flag == 1 {
            continue
        }
        track = append(track, num)
        ans = backtrack(nums, track, ans)
        track = track[:len(track) - 1]
    }
    return ans
}

File: leetcode/46.py
from typing import List


class Solution:
    def __init__(self) -> None:
        self.res = []

    def permute(self, nums: List[int]) -> List[List[int]]:
        li = []
        self.backtrack(nums, li)
        return self.res

    def backtrack(self, nums, track):
        size = len(nums)
        if size == len(track):
            temp = track.copy()
            self.res.append(temp)
            return
        for num in nums:
            if num in track:
                continue
            track.append(num)
            self.backtrack(nums, track)
            track.pop()


if __name__ == "__main__":
    res = Solution().permute([1, 2, 3])
    print(res)
